components/script/dom/globalscope.rs

```diff
@@ -3091,6 +3104,109 @@ impl GlobalScope {
             .handle_server_msg(scope, result);
     }
 
+    pub fn handle_gamepad_event(&self, gamepad_event: script_traits::GamepadEvent) {
+        match gamepad_event {
+            script_traits::GamepadEvent::Connected(index, name, bounds) => {
+                self.handle_gamepad_connect(
+                    index.0,
+                    name,
+                    bounds.axis_bounds,
+                    bounds.button_bounds,
+                );
+            },
+            script_traits::GamepadEvent::Disconnected(index) => {
+                self.handle_gamepad_disconnect(index.0);
+            },
+            script_traits::GamepadEvent::Updated(index, update_type) => {
+                self.handle_gamepad_update(index.0, update_type);
+            },
+        };
+    }
+
+    /// <https://www.w3.org/TR/gamepad/#dfn-gamepadconnected>
+    pub fn handle_gamepad_connect(
+        &self,
+        index: usize,
+        name: String,
+        axis_bounds: (f64, f64),
+        button_bounds: (f64, f64),
+    ) {
+        // TODO: 2. If document is not null and is not allowed to use the "gamepad" permission,
+        //          then abort these steps.
+        let this = Trusted::new(&*self);
+        self.gamepad_task_source().queue(
+            task!(gamepad_connected: move || {
+                let global = this.root();
+                let gamepad = Gamepad::new(&global, index as u32, name, axis_bounds, button_bounds);
+
+                // TODO: 3.4 If navigator.[[hasGamepadGesture]] is true:
+                // TODO: 3.4.1 Set gamepad.[[exposed]] to true.
+                if let Some(window) = global.downcast::<Window>() {
+                    let gamepad_list = window.Navigator().GetGamepads();
+                    let gamepad_arr: [DomRoot<Gamepad>; 1] = [gamepad.clone()];
+                    gamepad_list.add_if_not_exists(&gamepad_arr);
+
+                    if window.Document().is_fully_active() {
+                        gamepad.update_connected(true);
+                    }
+                }
+            }),
+            &self,
+        )
+        .unwrap();
+    }
+
+    /// <https://www.w3.org/TR/gamepad/#dfn-gamepaddisconnected>
+    pub fn handle_gamepad_disconnect(&self, index: usize) {
+        let this = Trusted::new(&*self);
+        self.gamepad_task_source()
+            .queue(
+                task!(gamepad_disconnected: move || {
+                    let global = this.root();
+                    if let Some(window) = global.downcast::<Window>() {
+                        let gamepad_list = window.Navigator().GetGamepads();
+                        if gamepad_list.Length() > 0 {
+                            gamepad_list.remove_gamepad(index);
```

Comment:
let's do:

1. if let some gamepad = gamepad_list.Item
2. update_connected(false)
3. remove_gamepad(which just removes it)